### DISCUSSION TITLE
webpack with esm libraries - watch mode warnings or errors?

### DISCUSSION THREAD

**Discussion by jeffrson** [webpack/webpack#16539] [Q&A]
Hi,
I'm converting our code base to ESM (currently "type" "module", webpack.config.cjs - source is .ts/.tsx/.scss).

Now this works fine with `webpack build`, however, for `webpack serve` (watch mode), there are a couple of disturbing "warnings", like this:
```
WARNING in ./src/root/root.tsx 18:32-43
export 'Application' (imported as 'Application') was not found in '@scope/library-name' (module has no exports)
 @ ./src/index.tsx 6:0-38 12:24-28
```

"library-name" definitely has an "exports" section - which is used in build mode.
So what could be the problem? Or is it indeed just a "warning"? What should I do about it?

Thanks
Jeff

**Comment by alexander-akait** [comment]
Sounds like you try to use `export` but it doesn't exists, need to look at code and configuration, thank you

**Comment by jeffrson** [comment]
Since this occurs in our mono-repo I need to extract some repro. Thank you so far.

**Comment by jeffrson** [comment]
While I was not quite true in the question (the WARNING occurs in build too, but had been suppressed by '--stats error-only') I made a repro at https://github.com/jeffrson/webpack_esm_test with two commits:
* commonjs - shows how it works with cjs (simple react "component")
* esm - shows warning
> export 'CustomHeadline' (imported as 'CustomHeadline') was not found in '@internal/lib-client' (module has no exports)

Moreover, the second commit has the problem that bothers me a lot more. "React" is undefined in the app (which then fails as a whole).

Any idea what I did wrong and how to fix it?

Thanks a lot
Jeff

**Reply by alexander-akait** [reply]
You have `"type": "module"`, but your code is CJS in `dist`, it is invalid, try to don't use webpack and just import your library using `Node.js` (you will see the error) - you can create `dist/cjs/index.js` and override `package.json`, webpack thinkgs your code in ESM, but your code is not in ESM

**Reply by jeffrson** [reply]
Thank you for looking at it!

The code in lib-client/dist was created by webpack as well (transpiled from ts-loader) - library target "umd". So should I use indeed library.type "module"? How safe (or stable) is it, given I have to opt-in via "experiments"?

I thought it would simply work as "umd" ("universal")...

**Reply by alexander-akait** [reply]
> How safe (or stable) is it, given I have to opt-in via "experiments"?

It is safe for most of cases, expect some edge cases, but you will see it while you development if you faced with such bugs (unlikely it will be).

UMD is such legacy things, and when you set `type: "module"` you are telling to Node.js - `my projects is ESM`, so everything will be considered from this side. I want to say it is not a ts/ts-loader/webpack bug. 

Even more - why do you bundle your library, because if you use only js/ts (and don't have side effects things like changing DOM, inject CSS and etc in top level scope) I recommend not to use bundler at all and keep your code in ESM, double bundling is lose many perfomance and your application can be bigger in this case

**Reply by jeffrson** [reply]
Well, yes, unfortunately there are <name>.module.scss files imported from typescript for the components in lib-client that are not found by application if not bundled.

**Reply by alexander-akait** [reply]
@jeffrson Just intresting why you want to bundle `CSS`, you `embedded` your styles wtihout possible to rewrite. 

If you consider the code as a library, then this is not a good idea, you don’t know where it will be used and developer might want to rewrite them and there are a lot of other cases when something in styles will be need to change. In that case better to complie CSS separately and provide example in docs how to use it in bundle (so developer can use an own library or rewrite them).

If you still want to embed your CSS (widgets/external embed), you can use vanially JS and DOM/CSSOM API and it will be more portable.

Note - the first thing allow to your lbirary will be more optimized (and better tree shaking and not only), using the second solution you can make pure function as `/* PURE */` (no side effects, i.e. all functions which change DOM/inject CSS are **not** `pure`) and have good optimization too.

**Reply by jeffrson** [reply]
Well, I'm (still) more the devops and backend guy. So, as far as I understand, our lib is rather "internally shared" code than a "general" library. (S)CSS is used to specify some pre-determined "defaults".

Of course you're right regarding flexibility and optimizations! Indeed, I would prefer if we could change the lib to get rid of webpack in the libraries. I will discuss it with our frontend team. Thanks a lot!!!

**Reply by alexander-akait** [reply]
Yeah, webpack is useful for bundling applications and optimize them, but now is 2022 and ESM works in most of modern browsers, most of libraries migrated (or start to migrate) to ESM, it will have many benefits, and therefore, when the developers builds their application, they themselves will decide which browsers and versions wnt to support and how inject dependecies and how optimize them, double bundling is not good idea at all, because you will lose a lot of optimizations
